### DISCUSSION TITLE
How to determine CLAP version from C(++) preprocessor?

### DISCUSSION THREAD

**Discussion by Arakula** [free-audio/clap#252] [Q&A] [answered]
In 1.1.4, some draft extensions were obviously dropped (file-reference.h, quick-controls.h).

There's a bit of a problem with that, if host and PlugIn versions differ and (at least) one of them supports one of the draft extensions.
Since I try to support as much as possible in my hosts, I'd like to keep that support intact for pre-1.1.4 PlugIns. Normally, in this case, I would simply copy the contents of 1.1.3 file-reference.h and quick-controls.h into my code and surround them with a little `#if`or `#ifdef` statement so that they are only used when needed ... but there's no way I can see to do that.

Unless I'm missing something, version.h doesn't include anything that can be used in a preprocessor statement to determine which CLAP version is in use. All nice and helpful for an `if()`statement in C(++), but not for the preprocessor.

Neither file-reference.h nor quick-controls.h contain any preprocessor definition, so I can't check for that either.

Of course, I _could_ simply copy the two headers from 1.1.3 to my copy of CLAP 1.1.4 and add them to clap.h, but I'd prefer to leave CLAP intact.
Of course, I _could_ simply copy the contents of the two headers to my CLAP glue code, shrug my shoulders thinking "Hey, I'll never step back to pre-1.1.4 anyway", and be done with it. I just don't like the idea.

Am I missing something here? Or can somebody show me how to cobble the required `#if` or `ifdef` statements together?

**Comment by baconpaul** [comment]
So the way JUCE does it is perhaps something we could follow. That is we can

```
#define CLAP_PREPROCESSOR_VERSION = ((CLAP_VERSION_MAJOR << 16) + (CLAP_VERSION_MINOR << 8) + CLAP_VERSION_REVISION)
```

then you can write things like

```
#if CLAP_PREPROCESSOR_VERSION < 0x010104

#else

#endif 
```

this exact JUCE_VERSINO strategy in juce headers allows us to make the clap juce extensions work with multiple juces

We may be hurt a bit by the clap versions having an explicit uint32_t in them so we need to be a bit careful

Thoughts?

**Reply by Arakula** [reply]
Here's a little sample program that shows that it unfortunately doesn't work in this simplicity.
This little program should compile as C or C++:

```
#include <stdint.h>
#include <stdio.h>

#if 0

#define CLAP_VERSION_MAJOR ((uint32_t)1)
#define CLAP_VERSION_MINOR ((uint32_t)1)
#define CLAP_VERSION_REVISION ((uint32_t)4)

#else

#define CLAP_VERSION_MAJOR 1
#define CLAP_VERSION_MINOR 1
#define CLAP_VERSION_REVISION 4

#endif

#define CLAP_PREPROCESSOR_VERSION ((CLAP_VERSION_MAJOR << 16) + (CLAP_VERSION_MINOR << 8) + CLAP_VERSION_REVISION)


#if CLAP_PREPROCESSOR_VERSION < 0x010105
#define CHK1 "<"
#else
#define CHK1 ">="
#endif

#if CLAP_PREPROCESSOR_VERSION > 0x010103
#define CHK2 ">"
#else
#define CHK2 "<="
#endif

#if CLAP_PREPROCESSOR_VERSION == 0x010104
#define CHK3 "=="
#else
#define CHK3 "!="
#endif


int main()
{
	printf("%d.%d.%d %s 1.1.5\n", CLAP_VERSION_MAJOR, CLAP_VERSION_MINOR, CLAP_VERSION_REVISION, CHK1);
	printf("%d.%d.%d %s 1.1.3\n", CLAP_VERSION_MAJOR, CLAP_VERSION_MINOR, CLAP_VERSION_REVISION, CHK2);
	printf("%d.%d.%d %s 1.1.4\n", CLAP_VERSION_MAJOR, CLAP_VERSION_MINOR, CLAP_VERSION_REVISION, CHK3);

	return 0;
}
```

If CLAP_VERSION_MAJOR/MINOR/REVISION were simple integer constants, it would work. In version.h, however, they are not - they are ((uint32_t)something). That's a bit too much for the preprocessor. Just change the `#if 0`  to `#if 1` in the 4th line and see for yourself what happens.
Tried with VS2022; even with /std:c++20 it doesn't work.

**Reply by baconpaul** [reply]
Right so then the trick is 

have a clap major version digit 
Clap major version references that with the uint
Then add the shifted combo using digit 

right?

@abique i think we should at least consider that for 1.1.5 (which is imminent) and then @Arakula can skip 114 and use the digit defined as 113 or earlier condition if needs be

**Reply by Arakula** [reply]
The method itself is tried and proven; Steinberg does it similarly:

https://github.com/steinbergmedia/vst3_pluginterfaces/blob/bc5ff0f87aaa3cd28c114810f4f03c384421ad2c/vst/vsttypes.h

**Reply by baconpaul** [reply]
yup i think #254 would do it also just doing some review work on it and fixing some of my silly bugs. Also some folks parse the .h file textually with custom software so theres one or two things we need to fix up.

**Comment by baconpaul** [comment]
(It also limits us to 256 makjor, minor or revision versions of CLAP before each breaks, so we couldn't do 1.13.257)

**Comment by baconpaul** [comment, marked as answer]
https://github.com/free-audio/clap/pull/254 this is roughly what it would look like if we changed this in 1.1.5 I think (but obviously that conflicts with next)

**Reply by Arakula** [reply]
Hm. So ... how exactly is CLAP_VERSION_LT() supposed to work?

Because ... it doesn't work as I would expect.
I just tried it with Clap 1.1.5. With that, I would expect `CLAP_VERSION_LT(1,1,4)` to evaluate to `false`, based on my idea that this macro invocation should mean "Is the current CLAP version less than 1.1.4".

Yeah well, not so.
Using CLAP 1.1.5,
```
#if CLAP_VERSION_LT(1,1,4)
#error CLAP version is lower than 1.1.4!
#endif
```
results in the error.
Logical, if I reduce `CLAP_VERSION_LT(1,1,4)` by hand:
```
#define CLAP_VERSION_LT(maj,min,rev) (((maj) < CLAP_VERSION_MAJOR) || \
                    ((maj) == CLAP_VERSION_MAJOR && (min) < CLAP_VERSION_MINOR ) || \
                    ((maj) == CLAP_VERSION_MAJOR && (min) == CLAP_VERSION_MINOR && (rev) < CLAP_VERSION_REVISION))
```
-->
```
CLAP_VERSION_LT(1,1,4) ((1 < 1) || \                 // nope
                        (1 == 1 && 1 < 1) || \       // nope
                        (1 == 1 && 1 == 1 && 4 < 5)) // YES!
```
So ... what's wrong: the macro or my interpretation? I can easily live with the second, although it seems counterintuitive to have to write
```
#if CLAP_VERSION_LT(1,1,4)
  // copy of the stuff that was part of the CLAP 1.1.3 draft extensions
#endif
```

**Reply by robbert-vdh** [reply]
Yeah seems like the macro performs the comparison the wrong way around.

**Reply by baconpaul** [reply]
argh.... the macro is named improperly. The macro implements "Is the given version in the argument less than the current version" but the name reads as "is the current clap version less than" which is the exact opposite. 1.1.4 is indeed less than 1.1.5. The correct code is

```
#if CLAP_VERSION_LT(1,1,4)
#error CLAP version 1.1.4 is lower than the current version!
#endif
```

@abique maybe we should do a quick 1.1.6 once we figure out what to name this thing. darnit.

I wish we had flagged this in the review.

**Reply by baconpaul** [reply]
We have two choices I think

1. swap the order in the macro and do a 1.1.6 (I think this is probably best) or
2. rename the macro to something else

I think 1 is best and I can prep that as a PR.

Sorry folks.

**Reply by robbert-vdh** [reply]
I think the 'is the current clap version less than' version has a lot more practical applications, and that's in line with how these kinds of version comparisons usually works. So my vote is for keeping the names as is and swapping the arguments. Noone's using these yet so that should be fine. :wink:

**Reply by baconpaul** [reply]
Yeah I'll prep that as a PR.

Sigh.

**Reply by baconpaul** [reply]
https://github.com/free-audio/clap/pull/269

@abique @robbert-vdh can you peek?

**Reply by baconpaul** [reply]
Anyway @Arakula this is fixed in 1.1.6, which has the only difference between it and 1.1.5 correcting that macro.

Thanks

**Reply by Arakula** [reply]
Thank you!
